src/common/base/src/base/progress.rs

```diff
@@ -24,24 +24,43 @@ pub struct ProgressValues {
     pub bytes: usize,
 }
 
+/// [`ProgressHook`] can be used to hook the progress to update the global metrics
+/// whenever the progress is updated.
+pub trait ProgressHook: std::fmt::Debug + Send + Sync {
+    fn incr(&self, progress_values: &ProgressValues);
+}
+
 #[derive(Debug)]
 pub struct Progress {
     rows: AtomicUsize,
     bytes: AtomicUsize,
+    hook: Option<Box<dyn ProgressHook>>,
 }
 
 impl Progress {
     pub fn create() -> Self {
         Self {
             rows: AtomicUsize::new(0),
             bytes: AtomicUsize::new(0),
+            hook: None,
+        }
+    }
+
+    pub fn with_hook(self, hook: Box<dyn ProgressHook>) -> Self {
+        Self {
+            rows: self.rows,
+            bytes: self.bytes,
+            hook: Some(hook),
         }
     }
 
     pub fn incr(&self, progress_values: &ProgressValues) {
         self.rows.fetch_add(progress_values.rows, Ordering::Relaxed);
         self.bytes
             .fetch_add(progress_values.bytes, Ordering::Relaxed);
+        if let Some(hook) = &self.hook {
```

Comment:
this might result in double counting if cluster mode.

Reply:
can you help explain more about this?